Fix Timeseries/recent endpoint speed

rohaan-rma opened the pull request:
Fixes #1078 
Proposed query update that changes a single Timeseries/Recent request from 30 seconds down to half a second on our SWT db export.

```sql
WITH base_ids AS (
    SELECT DISTINCT
        ( ts_code ),
        cwms_ts_id
    FROM
        av_cwms_ts_id2
    WHERE
        cwms_ts_id IN ( 'HAMM.Stage-Alt.Inst.15Minutes.0.Decodes-Raw', 'HAMM.Stage.Inst.15Minutes.0.Decodes-Raw' )
), av_tsv_limited AS (
    SELECT
        *
    FROM
        at_tsv_2023
    WHERE
        date_time BETWEEN DATE '2024-06-24' AND DATE '2024-07-22'
        AND ts_code IN (
            SELECT
                ts_code
            FROM
                base_ids
        )
    UNION ALL
    SELECT
        *
    FROM
        at_tsv_2024
    WHERE
        date_time BETWEEN DATE '2024-06-24' AND DATE '2024-07-22'
        AND ts_code IN (
            SELECT
                ts_code
            FROM
                base_ids
        )
), max_values AS (
    SELECT
        b.cwms_ts_id,
        t.ts_code,
        t.date_time,
        t.value,
        t.version_date,
        t.data_entry_date,
        t.quality_code,
        ex.earliest_entry_time       start_date,
        ex.latest_entry_time         end_date,
        MAX(t.date_time)
        OVER(PARTITION BY t.ts_code) AS max_date_time
    FROM
             av_tsv_limited t
        JOIN base_ids      b ON b.ts_code = t.ts_code
        JOIN at_ts_extents ex ON ex.ts_code = t.ts_code
                                 AND ex.version_time = t.version_date
)
SELECT
    max_values.cwms_ts_id,
    max_values.date_time,
    max_values.version_date,
    max_values.data_entry_date,
    max_values.quality_code,
    max_values.start_date,
    max_values.end_date,
    cwms_util.get_default_units(
        cwms_ts.get_base_parameter_id(max_values.ts_code),
        'EN'
    )                    AS def_units,
    max_values.date_time AS max_date_time,
    cwms_util.convert_units(max_values.value,
                            cwms_util.get_default_units(
                      cwms_ts.get_base_parameter_id(max_values.ts_code),
                      'SI'
                  ),
                            cwms_util.get_default_units(
                      cwms_ts.get_base_parameter_id(max_values.ts_code),
                      'EN'
                  ))                   AS value_at_max_date
FROM
    max_values
WHERE
    max_values.date_time = max_values.max_date_time;
```

rohaan-rma committed:
Add office-id param to FindMostRecentsInRange, use as another filter

rohaan-rma committed:
Make whereCondition field separate since office may be null

rohaan-rma committed:
Merge branch 'develop' of https://github.com/USACE/cwms-data-api into timeseries-endpoint-speed

rohaan-rma committed:
Work in Progress on JOOQ for new query, added in lots of boilerplate, still updating individual pieces

rohaan-rma committed:
Added most of the new query logic using JOOQ

rohaan-rma committed:
Add JOOQ query except final query fetch

rohaan-rma committed:
Merge branch 'develop' into timeseries-endpoint-speed

rohaan-rma committed:
Fix date imports

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/data/dao/TimeSeriesDaoImpl.java:
```diff
@@ -915,82 +898,122 @@ private Tsv buildTsvFromViewRow(usace.cwms.db.jooq.codegen.tables.records.AV_TSV
+                    .from(baseIds);
+
+            // references to appropriate year tables
+            Table<?> AT_TSV_2023_TABLE = table(name("AT_TSV_2023"));
```
These should be AT_TSV_PREV_YEAR and AT_TSV_CURRENT_YEAR and calculated at usage. Otherwise we'll have to remember to update them constantly.

rohaan-rma replied:
Yup you're correct. I just had those as place holders for now.

MikeNeilson reviewed (commented):
Looks reasonable. Definitely need to chew on it a bit. Fairly complex query, though I doubt unnecessarily so.

rohaan-rma committed:
Fix field accessing issue

rohaan-rma committed:
Merge branch 'develop' into timeseries-endpoint-speed

rohaan-rma committed:
Add new query, fix year tables, formatting

rohaan-rma committed:
Merge branch 'develop' into timeseries-endpoint-speed

rohaan-rma committed:
Fix imports

rohaan-rma committed:
Fix unit conversion in new query

rohaan-rma committed:
Clean up unnecessary helper methods with only one usage

rohaan-rma committed:
Add office-id to where condition

rohaan-rma commented:
Newly fixed proposed DB query:
- Fixed unit conversion (Added unit_id from AV_CWMS_TS_ID2)
- Added office-id to the where condition (Adds office-id only if not null)
```sql
with
  "base_ids" as (
    select distinct "CWMS_20"."AV_CWMS_TS_ID2"."TS_CODE", "CWMS_20"."AV_CWMS_TS_ID2"."CWMS_TS_ID", "CWMS_20"."AV_CWMS_TS_ID2"."UNIT_ID"
    from "CWMS_20"."AV_CWMS_TS_ID2"
    where (
      "CWMS_20"."AV_CWMS_TS_ID2"."CWMS_TS_ID" in ('HAMM.Stage.Inst.15Minutes.0.Decodes-Raw')
      and "CWMS_20"."AV_CWMS_TS_ID2"."DB_OFFICE_ID" = 'SWT'
    )
  ),
  "tsv_limited" as (
    select *
    from "CWMS_20"."AT_TSV_2023"
    where (
      "AT_TSV_2023"."DATE_TIME" between date '2024-06-24' and date '2024-07-22'
      and "AT_TSV_2023"."TS_CODE" in (
        select "base_ids"."TS_CODE"
        from "base_ids"
      )
    )
    union all
    select *
    from "CWMS_20"."AT_TSV_2024"
    where (
      "AT_TSV_2024"."DATE_TIME" between date '2024-06-24' and date '2024-07-22'
      and "AT_TSV_2024"."TS_CODE" in (
        select "base_ids"."TS_CODE"
        from "base_ids"
      )
    )
  ),
  "max_values" as (
    select
      "base_ids"."CWMS_TS_ID",
      "base_ids"."UNIT_ID",
      "tsv_limited"."TS_CODE",
      "tsv_limited"."DATE_TIME",
      "tsv_limited"."VALUE",
      "tsv_limited"."VERSION_DATE",
      "tsv_limited"."DATA_ENTRY_DATE",
      "tsv_limited"."QUALITY_CODE",
      "CWMS_20"."AT_TS_EXTENTS"."EARLIEST_ENTRY_TIME" "START_DATE",
      "CWMS_20"."AT_TS_EXTENTS"."LATEST_ENTRY_TIME" "END_DATE",
      max("tsv_limited"."DATE_TIME") over (partition by "tsv_limited"."TS_CODE") "max_date_time"
    from "tsv_limited"
      join "base_ids"
        on "base_ids"."TS_CODE" = "tsv_limited"."TS_CODE"
      join "CWMS_20"."AT_TS_EXTENTS"
        on (
          "CWMS_20"."AT_TS_EXTENTS"."TS_CODE" = "tsv_limited"."TS_CODE"
          and "CWMS_20"."AT_TS_EXTENTS"."VERSION_TIME" = "tsv_limited"."VERSION_DATE"
        )
  )
select
  "max_values"."CWMS_TS_ID",
  "max_values"."DATE_TIME",
  "max_values"."VERSION_DATE",
  "max_values"."DATA_ENTRY_DATE",
  "max_values"."QUALITY_CODE",
  "max_values"."START_DATE",
  "max_values"."END_DATE",
  "CWMS_20"."CWMS_UTIL"."GET_DEFAULT_UNITS"(
    "CWMS_20"."CWMS_TS"."GET_BASE_PARAMETER_ID"("max_values"."TS_CODE"),
    'EN'
  ) "def_units",
  "max_values"."DATE_TIME" "max_date_time",
  "CWMS_20"."CWMS_UTIL"."CONVERT_UNITS"(
    "max_values"."VALUE",
    "max_values"."UNIT_ID",
    "CWMS_20"."CWMS_UTIL"."GET_DEFAULT_UNITS"(
      "CWMS_20"."CWMS_TS"."GET_BASE_PARAMETER_ID"("max_values"."TS_CODE"),
      'EN'
    )
  ) "value_at_max_date"
from "max_values"
where "max_values"."DATE_TIME" = "max_values"."max_date_time"
```

rohaan-rma committed:
Add missing whereCondition assignment for office-id

rohaan-rma committed:
Fix category and group recent endpoint

rohaan-rma commented:
Old Query with over 50s runtime.
<img width="2401" height="1448" alt="Screenshot 2025-07-22 115911" src="https://github.com/user-attachments/assets/a1596e99-0dc2-4c21-b0d3-3323da45a026" />

New and Improved query with under 1s runtime.
<img width="2334" height="1249" alt="Screenshot 2025-07-22 115543" src="https://github.com/user-attachments/assets/b65a251c-f138-4755-bb1c-bb9c3b72c321" />

rohaan-rma committed:
Merge branch 'develop' into timeseries-endpoint-speed

rohaan-rma commented:
Integration Tests to show improved query time.

Old Query:
<img width="1822" height="1404" alt="Screenshot 2025-07-23 104401" src="https://github.com/user-attachments/assets/f3f456f1-30a1-4451-91f7-abab26b56803" />

New Query:
<img width="2217" height="1424" alt="Screenshot 2025-07-23 103544" src="https://github.com/user-attachments/assets/a1750431-cd9a-4f4d-804b-96d78da97569" />

MikeNeilson commented:
Excellent Set that test up with a reasonable time assert, I think there's an existing check int the assertion. If not, and maybe there isn't since I wrote this... copy this over: https://github.com/opendcs/opendcs/blob/main/integrationtesting/fixtures/src/main/java/org/opendcs/fixtures/assertions/Waiting.java

You're local performance probably isn't the right target, but something like 5 seconds, just to make sure we aren't causing performance regressions later on.

rohaan-rma committed:
Add conditional for tsv_limited tables union, and Update category/group processing to match new query

rohaan-rma committed:
Add select distinct to avoid duplicate rows being returned

rohaan-rma committed:
Clean code

rohaan-rma commented:
A few changes: 
- Only uses one table for tsv_limited if both start and end date are in the same year
- Updated category/group processing in the timeseries/recent endpoint to use optimized query

Old query:
<img width="1809" height="1853" alt="Screenshot 2025-07-23 150057" src="https://github.com/user-attachments/assets/894bc2e6-3e45-42cf-bb4c-086dd7161490" />
New query:
<img width="1729" height="1547" alt="Screenshot 2025-07-23 144012" src="https://github.com/user-attachments/assets/f1bc0a6c-2023-4654-bc42-9702c6460874" />

Note: These integration tests are only to show time improvement from the old query to the new query. These tests are using the jdbc export and are not intended to be pushed anywhere. I am considering adding the .time(lessthan(5000 ms)) to the assertions in the current integration tests (Although the current IT aren't made for using with the export, so they would still pass under the time constraint because of a lack of populated data).

rohaan-rma committed:
Merge branch 'develop' into timeseries-endpoint-speed

rma-rripken reviewed (commented):
Looks good to me

rohaan-rma committed:
Add assertion for timing in IT

rohaan-rma committed:
Merge branch 'develop' into timeseries-endpoint-speed

rma-rripken commented on cwms-data-api/src/main/java/cwms/cda/data/dao/TimeSeriesDaoImpl.java:
```diff
@@ -349,9 +341,10 @@ protected TimeSeries getRequestedTimeSeries(String page, int pageSize, @NotNull
+        Field<Timestamp> dateTimeCol = field("DATE_TIME", Timestamp.class).as("DATE_TIME");
+        Field<Double> valueCol = field("VALUE", Double.class).as("VALUE");
+        Field<Integer> qualityCol = field("QUALITY_CODE", Integer.class).as("QUALITY_CODE");
+        Field<Timestamp> dataEntryDate = field("DATA_ENTRY_DATE", Timestamp.class).as("data_entry_date");
```
This should probably use the constants you added.

rohaan-rma replied:
Fixed. Sorry, I thought I had done that originally, but maybe after I resolved the merge conflicts this morning I might've accidentally changed them back.

rohaan-rma committed:
Continue using String constants
